### PR TITLE
[feature] implementation of nodegroup and edgeapplication controller 

### PR DESCRIPTION
**Pull request by Congrool**
**What type of PR is this?**
/kind feature
/kind api-change

**What this PR does / why we need it**:
Implementation of feature at #3582.
This pr includes 
1. NodeGroup CRD types, yaml and generated code using code-generator
2. NodeGroup controller

This controller is implemented through `controller-runtime@v0.10.3`, it will watch node and nodegroup resources and do the reconcilation as follows:

1. check if this nodegroup has been deleted, if so, remove `belonging-to` label on all its contained nodes and remove the finalizer.
2. check if this nodegroup has a finalizer, if not, add the finalizer to it.
3. add `belonging-to` label to new added nodes
4. remove `belonging-to` label on deleted nodes
5. update status of nodegroup

node add/update/delete event will be translated to nodegroup event and reconcile as described above.

2022.4.6 CHANGELOG:
1. add edgeapplication api and its generated code
2. add edgeapplication controller, including status manager and overrider manager.
- edgeapplication controller: main reconcile process
- status manager: manage the status field of edgeapplication api
- overrider manager: contains name overrider, replicas overrider and image overrider(from karmada), takes the responsibility of overriding the provided manifests.
3. add controller manager cmd: a new cmd to start nodegroup controller and edgeapplication controller.

/cc  @vincentgoat @fisherxu @lonelyCZ

### CODE REVIEW

File: build/crds/grouping/grouping_v1alpha1_nodegroup.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.6.2
+  creationTimestamp: null
+  name: nodegroups.grouping.kubeedge.io
+spec:
+  group: grouping.kubeedge.io
```
Comment (fisherxu) [inline build/crds/grouping/grouping_v1alpha1_nodegroup.yaml:11]: The group name needs more discussions, it's more related to apps.

Comment (fisherxu) [inline build/crds/grouping/grouping_v1alpha1_nodegroup.yaml:11]: The group name needs more discussions, it's more related to apps.

Comment (Congrool) [reply build/crds/grouping/grouping_v1alpha1_nodegroup.yaml:11]: 🤔 how about `appgroupings`?

**Comment by Congrool** [comment]
It has not been tested.
/hold

**Comment by vincentgoat** [comment]
[TEST CASE](https://docs.google.com/spreadsheets/d/1ngL5OFiBHwn35Ceo3p17DMqnaLFygMV8BvlMInv1yJw/edit#gid=0)

[NODEGROUP/EDGEAPPLICATION BEST PRACTICE](https://docs.google.com/document/d/1lzVNGPowng29yfUZFJ66epDhymX2ZZPkYpu0uqwZPQQ/edit#)

**Comment by vincentgoat** [comment]
> It has not been tested. /hold

Already done test @Congrool

**Comment by Congrool** [comment]
Thanks.

/unhold

File: pkg/apis/apps/v1alpha1/nodegroup_types.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+)
+
+// SelectionStatus represents the status of selecting a node as a member of this NodeGroup.
+type SelectionStatus string
+
+const (
+	// SucceededSelection represents that this node has been selected as a member of this NodeGroup.
+	SucceededSelection SelectionStatus = "Succeeded"
+	// FailedSelection represents that this node failed to become a member of this NodeGroup.
+	FailedSelection SelectionStatus = "Failed"
```
Comment (fisherxu) [inline pkg/apis/apps/v1alpha1/nodegroup_types.go:80]: If the node hasn't been selected, it should not be included in the status?

Comment (Congrool) [reply pkg/apis/apps/v1alpha1/nodegroup_types.go:80]: Yes, only nodes selected by node name or labels will have their status in the nodegroup, and the statuses of these nodes have two situations, including `SucceededSelection` and `FailedSelection`.  

Nodes that not selected by the nodegroup will not have their status entries.

Comment (fisherxu) [reply pkg/apis/apps/v1alpha1/nodegroup_types.go:80]: I'm not sure what's the meaning of `FailedSelection`? The node in nodegroup, but select failed?

Comment (Congrool) [reply pkg/apis/apps/v1alpha1/nodegroup_types.go:80]: Currently there're two situations that will have `FailedSelection`:
1. Use node name to select the node, but the node does not exist. In this situation, we should notify the users.
2. Select a node that already belongs to another nodegroup. We should make sure that one node can only belong to one nodegroup, so when user want to select a node that already blongs to another nodegroup, due to negligence or something else, we also should notify the users.

Comment (fisherxu) [reply pkg/apis/apps/v1alpha1/nodegroup_types.go:80]: Okey, sounds good, maybe we can add some comments later :)

Comment (Congrool) [reply pkg/apis/apps/v1alpha1/nodegroup_types.go:80]: Ok, I'll add later.

**Review by fisherxu — COMMENTED** [review]
/lgtm

**Pull request merged**